File: client/windows/ClientCompatibility.cs
namespace Codmes.Windows;

public static class ClientCompatibility
{
    public static bool Supports(
        IEnumerable<string> declaredPlatforms,
        IEnumerable<string> declaredFormFactors,
        string currentPlatform,
        string currentFormFactor)
    {
        var originalPlatforms = declaredPlatforms.Select(Normalize).Where(value => value.Length > 0).Distinct().ToArray();
        var platforms = originalPlatforms.Select(value => value == "ipados" ? "ios" : value).Distinct().ToArray();
        var declaredFactors = declaredFormFactors.Select(Normalize).Where(value => value.Length > 0).Distinct().ToArray();
        var factors = declaredFactors.Length > 0 ? declaredFactors : LegacyFormFactors(originalPlatforms);

        return (platforms.Length == 0 || platforms.Contains(Normalize(currentPlatform)))
            && (factors.Length == 0 || factors.Contains(Normalize(currentFormFactor)));
    }

    private static string[] LegacyFormFactors(IEnumerable<string> platforms) =>
        platforms.Select(value => value switch
        {
            "macos" => "desktop",
            "ios" => "phone",
            "ipados" => "tablet",
            _ => ""
        }).Where(value => value.Length > 0).Distinct().ToArray();

    private static string Normalize(string value) => value.Trim().ToLowerInvariant();
}

File: client/android/app/src/main/java/com/codmes/android/ClientCompatibility.kt
package com.codmes.android

object ClientCompatibility {
    fun supports(
        declaredPlatforms: List<String>,
        declaredFormFactors: List<String>,
        currentPlatform: String,
        currentFormFactor: String
    ): Boolean {
        val normalizedPlatforms = declaredPlatforms
            .map(::normalize)
            .map { if (it == "ipados") "ios" else it }
            .distinct()
        val normalizedFormFactors = if (declaredFormFactors.isNotEmpty()) {
            declaredFormFactors.map(::normalize).distinct()
        } else {
            legacyFormFactors(declaredPlatforms)
        }
        return (normalizedPlatforms.isEmpty() || normalize(currentPlatform) in normalizedPlatforms) &&
            (normalizedFormFactors.isEmpty() || normalize(currentFormFactor) in normalizedFormFactors)
    }

    private fun legacyFormFactors(platforms: List<String>): List<String> =
        platforms.mapNotNull {
            when (normalize(it)) {
                "macos" -> "desktop"
                "ios" -> "phone"
                "ipados" -> "tablet"
                else -> null
            }
        }.distinct()

    private fun normalize(value: String) = value.trim().lowercase()
}
